R/MSnExperiment-class.R

```diff
@@ -479,53 +496,84 @@ readMSnExperiment <- function(file, sampleData, backend = BackendMzR(),
-              object@backend <- backend
-              validObject(object)
-              object
-          })
-#' @rdname hidden_aliases
-setMethod("setBackend", c("MSnExperiment", "BackendMzR"),
-          function(object, backend, ..., BPPARAM = bpparam()) {
-              if (any(backend@modCount > 0))
-                  stop("Can not change backend to 'BackendMzR' because the ",
-                       "data was changed.")
-              object@backend <- backendInitialize(backend, fileNames(object),
-                                                  object@spectraData, ...)
-              validObject(object)
-              object
-          })
+setMethod(
+    "setBackend",
+    c("MSnExperiment", "Backend"),
+    function(object, backend, ..., BPPARAM = bpparam()) {
+    backend <- backendInitialize(backend, fileNames(object), object@spectraData,
+                                 ...)
+    ## update fileIdx, useful to split src backends across cores
+    spd <- object@spectraData
+    spd$fileIdx <- 1L
+    spd <- split(spd, object@spectraData$fileIdx)
+
+    backendSplitByFile(backend, object@spectraData) <-
+        bpmapply(function(dst, src, spd, queue) {
+            backendWriteSpectra(dst, backendReadSpectra(src, spd), spd)
+        },
+        dst = backendSplitByFile(backend, object@spectraData),
+        src = backendSplitByFile(object@backend, object@spectraData),
+        spd = spd,
+        SIMPLIFY = FALSE, USE.NAMES = FALSE, BPPARAM = BPPARAM)
+
+    object@backend <- backend
+    validObject(object)
+    object
+})
+
 #' @rdname hidden_aliases
-setMethod("setBackend", c("MSnExperiment", "BackendHdf5"),
-          function(object, backend, ..., BPPARAM = bpparam()) {
-              backend <- backendInitialize(backend, fileNames(object),
-                                           object@spectraData, ...)
-              spd <- split(object@spectraData, object@spectraData$fileIdx)
-              cnts <- bplapply(spd, function(z, hdf5_backend, backend) {
-                  res <- backendWriteSpectra(hdf5_backend,
-                                             backendReadSpectra(backend, z), z)
-                  res@modCount[z$fileIdx[1]]
-              }, hdf5_backend = backend, backend = object@backend,
-              BPPARAM = BPPARAM)
-              backend@modCount <- unlist(cnts)
-              object@backend <- backend
-              validObject(object)
-              object
+setMethod(
+    "setBackend",
+    c("MSnExperiment", "BackendMzR"),
+    function(object, backend, ..., BPPARAM = bpparam()) {
+    if (any(object@backend@modCount > 0))
+        stop("Can not change backend to 'BackendMzR' because the ",
+             "data was changed.")
+    object@backend <- backendInitialize(backend, fileNames(object),
+                                        object@spectraData, ...)
+    validObject(object)
+    object
 })
 
+#' @rdname MSnExperiment
+applyProcessingQueue <- function(x, BPPARAM = bpparam()) {
```

Comment:
BTW: I know we had this discussion before but is there a specific reason why this is a function and not a method?